util/httputil/client.go

```diff
@@ -0,0 +1,84 @@
+// Copyright 2013 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package httputil
+
+import (
+	"net"
+	"net/http"
+	"time"
+)
+
+func NewClient(rt http.RoundTripper) *http.Client {
+	return &http.Client{Transport: rt}
+}
+
+// NewDeadlineClient returns a new http.Client which will time out long running
+// requests.
+func NewDeadlineClient(timeout time.Duration) *http.Client {
+	return NewClient(NewDeadlineRoundTripper(timeout))
+}
+
+func NewDeadlineRoundTripper(timeout time.Duration) http.RoundTripper {
+	return &http.Transport{
+		// We need to disable keepalive, because we set a deadline on the
+		// underlying connection.
+		DisableKeepAlives: true,
+		Dial: func(netw, addr string) (c net.Conn, err error) {
+			start := time.Now()
+
+			c, err = net.DialTimeout(netw, addr, timeout)
+
+			if err == nil {
+				c.SetDeadline(start.Add(timeout))
+			}
+
+			return
+		},
+	}
+}
+
+type bearerAuthRoundTripper struct {
+	bearerToken string
+	rt          http.RoundTripper
+}
+
+// NewBearerAuthRoundTripper adds the provided bearer token to a request unless the authorization
+// header has already been set.
+func NewBearerAuthRoundTripper(bearer string, rt http.RoundTripper) http.RoundTripper {
+	return &bearerAuthRoundTripper{bearer, rt}
+}
+
+func (rt *bearerAuthRoundTripper) RoundTrip(req *http.Request) (*http.Response, error) {
+	if len(req.Header.Get("Authorization")) != 0 {
```

Comment:
As the return line is the same, I'd invert the condition and set the header on match.
Just what feels more intuitive to me, though.